File: python/maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        prev = None
        runner = head
        current = head
        while runner:
            runner = runner.next.next
            aux = current.next
            current.next = prev
            prev = current
            current = aux
        maximal = 0
        while current:
            maximal = max(maximal, prev.val + current.val)
            prev = prev.next
            current = current.next
        return maximal

File: go/maximum-twin-sum-of-a-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */


 func pairSum(head *ListNode) int {
    current := head
    runner := head
    var prev *ListNode = nil
    for runner != nil {
        runner = runner.Next.Next
        aux := current.Next
        current.Next = prev
        prev = current
        current = aux
    }
    maximal := 0
    for current != nil {
        sum := current.Val + prev.Val
        if maximal < sum {
            maximal = sum
        }
        current = current.Next
        prev = prev.Next
    }
    return maximal
}
